transports/tcp/src/lib.rs

```diff
@@ -270,73 +312,140 @@ impl Future for TcpDialFut {
 
     fn poll(&mut self) -> Poll<Option<Self::Item>, io::Error> {
-        if self.tell_new_addr {
-            self.tell_new_addr = false;
-            return Ok(Async::Ready(Some(ListenerEvent::NewAddress(self.listen_addr.clone()))))
-        }
-
         let inner = match self.inner {
             Ok(ref mut inc) => inc,
-            Err(ref mut err) => {
-                return Err(err.take().expect("poll called again after error"));
-            }
+            Err(ref mut err) => return Err(err.take().expect("poll called again after error"))
         };
 
         loop {
-            match inner.poll() {
-                Ok(Async::Ready(Some(sock))) => {
-                    let addr = match sock.peer_addr() {
-                        // TODO: remove this expect()
-                        Ok(addr) => addr
-                            .to_multiaddr()
-                            .expect("generating a multiaddr from a socket addr never fails"),
-                        Err(err) => {
-                            // If we can't get the address of the newly-opened socket, there's
-                            // nothing we can do except ignore this connection attempt.
-                            error!("Ignored incoming because could't determine its \
-                                    address: {:?}", err);
-                            continue
-                        },
-                    };
-
-                    match apply_config(&self.config, &sock) {
-                        Ok(()) => (),
-                        Err(err) => {
-                            let event = ListenerEvent::Upgrade {
-                                upgrade: future::err(err),
-                                listen_addr: self.listen_addr.clone(),
-                                remote_addr: addr
-                            };
-                            return Ok(Async::Ready(Some(event)))
+            if let Some(event) = self.pending.pop_front() {
+                return Ok(Async::Ready(Some(event)))
+            }
+
+            let sock = match inner.poll() {
+                Ok(Async::Ready(Some(sock))) => sock,
+                Ok(Async::Ready(None)) => return Ok(Async::Ready(None)),
+                Ok(Async::NotReady) => return Ok(Async::NotReady),
+                Err(()) => unreachable!("sleep_on_error never produces an error")
+            };
+
+            let sock_addr = match sock.peer_addr() {
+                Ok(addr) => addr,
+                Err(err) => {
+                    // If we can't get the address of the newly-opened socket, there's
+                    // nothing we can do except ignore this connection attempt.
+                    error!("Failed to get peer address: {:?}", err);
+                    continue
+                }
+            };
+
+            let listen_addr = match sock.local_addr() {
+                Ok(addr) => match self.addrs.as_mut() {
+                    Either::Left(ma) => ma.clone(),
```

Comment:
I'm not a fan of using `Left` and `Right` here instead of more explicit variants.

Reply:
Sorry, I did not expect your review so soon ;-). I had just force pushed to this branch.